File: python/Unidad_2/ej9.py
# Aplique a una imagen grayscale la transformación apropiada para que su 
# histograma se comprima a un cierto rango definido por el usuario (shrinking).

import numpy as np
import matplotlib.pyplot as plt
from skimage import img_as_float
from skimage.color import gray2rgb
from skimage.io import imread
from skimage.exposure import rescale_intensity
import tkinter as tk
from tkinter import filedialog

def plot_img_hist(img, num_plot, title, bins=255):
    '''
    Plots de imagen e histograma
    '''
    # Imagen
    plt.subplot(num_plot)
    if len(img.shape) == 3:
        plt.imshow(img, cmap='gray')
    else: # len(img.shape) == 2
        plt.imshow(gray2rgb(img), cmap='gray')
    plt.title(title)
    # Histograma
    plt.subplot(num_plot + 2)
    plt.hist(img_as_float(img).ravel(), bins=bins)
    plt.xlim(0, 1)
    
def file_read():
    file_path = filedialog.askopenfilename()
    return file_path


try:
    # Busco imagen y obtengo su ruta
    root = tk.Tk()
    root.withdraw() 
    img = imread(file_read())
    out_range = (0, 50)
    img_shrink = rescale_intensity(img, out_range=out_range).astype(np.uint8)
    plot_img_hist(img, 221, 'Imagen original')
    plot_img_hist(img_shrink, 222, 'Imagen shrinking', bins=out_range[1]-out_range[0])
    
    plt.show()
except:
    print('Cerraste la ventana!')

print('👋🏽')
root.destroy()

File: matlab/Unidad_2/ej9.m
% 9) Aplique a una imagen grayscale la transformación apropiada para que su
% histograma se comprima a un cierto rango definido por el usuario (shrinking).

[file, path] = uigetfile({'*.jpg';'*.JPG';'*.bmp';'*.tif'});
dir = strcat(path,file);
info = imfinfo(dir);

switch info.ColorType
    case 'grayscale'
        I=imread(dir);
    case 'truecolor'
        rgb=imread(dir);
        I=rgb2gray(rgb);
    case 'indexed'
        [X,map]=imread(dir);
        I=ind2gray(X,map);
end

I_min = min(I(:)); % Valor minimo de la imagen
I_max = max(I(:)); % Valor maxomo de la imagen
O_min = I_min/2; % Valor minimo final deseado
O_max = I_max/2; % Valor maximo final deseado

LH_in = [double(I_min)/255, double(I_max)/255]; % Low & High Input
LH_out = [double(O_min)/255, double(O_max)/255]; % Low & High Output
O = imadjust(I, LH_in, LH_out);

subplot(2, 2, 1);
imshow(I);
title('Imagen original');

subplot(2, 2, 2);
imshow(O);
title('Imagen con Shrinking');

subplot(2, 2, 3);
imhist(I);
title('Histograma original');

subplot(2, 2, 4);
imhist(O);
title('Histograma con Shrinking');